File: src/main/java/org/academy/internal/client/gui/imgui/ImGuiUtilInternal.java
package org.academy.internal.client.gui.imgui;

import com.mojang.blaze3d.pipeline.RenderTarget;
import com.mojang.blaze3d.systems.RenderSystem;
import imgui.ImGui;
import imgui.extension.implot.ImPlot;
import imgui.flag.ImGuiConfigFlags;
import imgui.gl3.ImGuiImplGl3;
import imgui.glfw.ImGuiImplGlfw;
import net.minecraft.client.Minecraft;
import org.jetbrains.annotations.ApiStatus;

import java.util.OptionalInt;

/**
 * 你必须通过 ImGuiUtilApi 间接调用喵, 不能直接使用喵
 */
@ApiStatus.Internal
public final class ImGuiUtilInternal {
    private static final ImGuiImplGl3 IM_GUI_IMPL_GL_3 = new ImGuiImplGl3();
    private static final ImGuiImplGlfw IM_GUI_IMPL_GLFW = new ImGuiImplGlfw();

    public static ImGuiImplGl3 getImGuiImplGl3() {
        return IM_GUI_IMPL_GL_3;
    }

    public static ImGuiImplGlfw getImGuiImplGlfw() {
        return IM_GUI_IMPL_GLFW;
    }

    public static void init() {
        var handle = Minecraft.getInstance().getWindow().handle();
        ImGui.createContext();
        ImPlot.createContext();

        var data = ImGui.getIO();
        data.setFontGlobalScale(1F);
        data.setConfigFlags(ImGuiConfigFlags.DockingEnable | ImGuiConfigFlags.NavEnableKeyboard);

        getImGuiImplGl3().init();
        getImGuiImplGlfw().init(handle, true);
    }

    public static void render(RenderTarget renderTarget, Runnable renderCommand) {
        var colorTextureView = renderTarget.getColorTextureView();
        if (colorTextureView == null) return;
        try (
                var ignored = RenderSystem.getDevice().createCommandEncoder().createRenderPass(
                        () -> "ImGui",
                        colorTextureView,
                        OptionalInt.empty()
                )
        ) {
            getImGuiImplGl3().newFrame();
            getImGuiImplGlfw().newFrame();
            ImGui.newFrame();

            renderCommand.run();

            ImGui.render();
            getImGuiImplGl3().renderDrawData(ImGui.getDrawData());
        }
    }

    public static void clearEventsQueue() {
        ImGui.getIO().clearEventsQueue();
    }

    public static void dispose() {
        getImGuiImplGl3().shutdown();
        getImGuiImplGlfw().shutdown();

        ImPlot.destroyContext();
        ImGui.destroyContext();
    }

    private ImGuiUtilInternal() {
    }
}

File: src/main/kotlin/org/academy/internal/client/gui/imgui/ImGuiUtilInternal.kt
package org.academy.internal.client.gui.imgui

import com.mojang.blaze3d.pipeline.RenderTarget
import net.minecraft.client.Minecraft
import org.jetbrains.annotations.ApiStatus

/**
 * 游戏内 ImGui 门面：把 [ImGuiBackend] 绑定到游戏窗口，保持原 [org.academy.api.client.gui.imgui.ImGuiUtilApi]
 * 行为不变。独立桌面编辑器直接使用 [ImGuiBackend]。
 */
@ApiStatus.Internal
object ImGuiUtilInternal {
    private var backend: ImGuiBackend? = null

    val imGuiImplGlfw: Any?
        get() = backend?.imGuiImplGlfw

    fun init() {
        val minecraft = Minecraft.getInstance()
        backend = ImGuiBackend(
            minecraft.window.handle()
        ).also { it.init() }
    }

    fun render(renderTarget: RenderTarget, renderCommand: () -> Unit) {
        backend?.render(renderTarget, renderCommand)
    }

    fun clearEventsQueue() {
        backend?.clearEventsQueue()
    }

    fun wantCaptureMouse(): Boolean = backend?.wantCaptureMouse() ?: false

    fun wantCaptureKeyboard(): Boolean = backend?.wantCaptureKeyboard() ?: false

    fun dispose() {
        backend?.dispose()
        backend = null
    }
}
